### PR TITLE
Run compiled methods from MethodHandle INLs 

### PR DESCRIPTION
**Pull request by gacholio**
If the target method of a MethodHandle INL invoked from compiled code has been compiled, the current code will run the method in the interpreter. This change will allow the compiled code to be executed.

Fixes: https://github.com/eclipse-openj9/openj9/issues/17321

Signed-off-by: Graham Chapman <graham_chapman@ca.ibm.com>

### CODE REVIEW

**Comment by gacholio** [comment]
jenkins test sanity zlinux jdk20

**Comment by gacholio** [comment]
jenkins test sanity zlinux jdk20

File: runtime/vm/BytecodeInterpreter.hpp [hunk 1]
```diff
@@ -620,15 +620,20 @@ class INTERPRETER_CLASS
...
-				if (J9_ARE_NO_BITS_SET(preCount, J9_STARTPC_NOT_TRANSLATED)) {
-					/* Already compiled */
+				if (startAddressIsCompiled(preCount)) {
+					if (immediatelyRunCompiledMethod) {
+						if (methodCanBeRunCompiled(_sendMethod)) {
+							rc = promotedMethodOnTransitionFromJIT(REGISTER_ARGS, (void*)_pc, (void*)preCount);
+							goto done;
+						}
+					}
 					break;
```
Comment (babsingh) [inline runtime/vm/BytecodeInterpreter.hpp:630]: We can compress the code by removing one of the `if` statements.
```suggestion
				if (immediatelyRunCompiledMethod && startAddressIsCompiled(preCount)) {
					if (methodCanBeRunCompiled(_sendMethod)) {
						rc = promotedMethodOnTransitionFromJIT(REGISTER_ARGS, (void*)_pc, (void*)preCount);
						goto done;
					}
					break;
```

**Review by babsingh — COMMENTED** [review]
Other than the minor nitpick, the code changes LGTM.

The old build should still be good after the above change is addressed: https://openj9-jenkins.osuosl.org/job/Pipeline_Build_Test_JDK20_s390x_linux/93/.

File: runtime/vm/BytecodeInterpreter.hpp (hunk 1 above)
Comment (gacholio) [reply runtime/vm/BytecodeInterpreter.hpp:630]: No, that code is incorrect - the compiled check must occur first.

**Review by babsingh — APPROVED** [review]

**Comment by mpirvu** [comment]
This change significantly improves throughput of a Liberty based application (RestCRUD) with JDK19 (~13%).
Since the change applies to all Java versions, does anyone have a feel for whether Java17 apps would benefit from this change as well? Thanks

**Comment by babsingh** [comment]
> Does anyone have a feel for whether Java17 apps would benefit from this change as well? 

Yes. This change is related to OJDK MHs which were enabled in OpenJ9 from JDK17 and onwards. Similar performance improvements should be seen in JDK17. JDK8 and JDK11 won't be impacted until OJDK MHs are back-ported to those JDK versions; currently, the back-port task is paused since all resources are dedicated to support the next releases.

**Comment by vijaysun-omr** [comment]
Thanks Babneet for answering the question about whether the fix also applies to JDK17. The magnitude of the performance issue that it causes may come down to the degree to which method handles are used in JDK 17 JCL (vs say how much they get used in JDK 19 JCL). So it would not surprise me if the impact was higher in JDK 19 (and future releases) since method handle use is only increasing in the JCL (is my sense at least).

**Pull request merged**